### ISSUE TITLE
Any tip for testing?

### ISSUE THREAD

**Issue by julian-linux** [react-hook-form/react-hook-form#532]
I'm using react-scripts  and material-ui and my code is something like
```
const Form = ({onSubmit}) => {
 const { register, handleSubmit, errors } = useForm()
 return (
  <form onSubmit={handleSubmit(onSubmit)}>
     <TextField {...} inputRef={register} />
     .....
      <Button type="submit" />
  </form>
}
```

i want to test the form doing something like 
```
describe('Form Component', () => {
 it('should submit when data filled', () => {
    const onSubmit = jest.fn()
    const component = mount(<Form onSubmit={onSubmit} />)
    const input = component.find(TextField).at(1)
    input.simulate('change', { target: { value: 'test'} })
    component.find(Button).first().simulate('submit')
    expect(onSubmit).toBeCalled()
 })
})
```

this throw me 2 errors: 
 1. onSubmit is not called
 2.
```
 Attempted to log "Warning: An update to Form inside a test was not wrapped in act(...).
  ....

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
....
```
i put every `simulate` in an `act` function and does not work.

Any tip is welcome

**Comment by bluebill1049** [comment]
Can you please post this question in our spectrum channel? Some of the community member may be able to help u out there.

**Comment by bluebill1049** [comment]
Try async 

```
    act(aysnc () => {

    });
```

**Comment by markusschlebusch** [comment]
Just had the same issuebut I solved it with the following code:

`
    act(async () => {
      .....
      wrapper.find('form').simulate('submit');
      expect().toHaveBeenCalledTimes(1);
    });
`
When the expect is after the act()-part it doesn't work because of a timing issue.

**Comment by bluebill1049** [comment]
Thanks @mschlebusch !

**Comment by mschlebusch** [comment]
Sorry this solution didn't work. Had some confusing issues locally.

**Comment by bluebill1049** [comment]
what's the confusing issue? @mschlebusch

**Comment by bluebill1049** [comment]
I will recommend you guys to use test-library with hooks. That's what we using in this lib.

**Comment by bluebill1049** [comment]
post your question here: https://spectrum.chat/react-hook-form?tab=posts our community can help out too.

**Comment by julian-linux** [comment]
I have a solution.
I had to change to `react-testing-library` for this. Enzyme does not work as expected.

This is an idea of my form.... and i added some props for `react-testing-library`
```
const Form = ({onSubmit}) => {
 const { register, handleSubmit, errors } = useForm()
 return (
  <form onSubmit={handleSubmit(onSubmit)} data-testid='form'>
     <TextField 
        {...}
        inputRef={register} 
        inputProps={{
          'aria-label': 'input'
        }}/>
     .....
      <Button type="submit" />
  </form>
}
```
And this is an idea of my new test
```
describe('Form Component', () => {
 it('should submit when data filled', async () => {
  const onSubmit = jest.fn()
  const { getByLabelText, getByTestId } = render(<Form onSubmit={onSubmit} />)
  const input = getByLabelText('input')
  await act(async () => {
    await fireEvent.change(input, { target: { value: '123' } })
  })
  await act(async () => {
      fireEvent.submit(getByTestId('form'))
  })
  expect(onSubmit).toBeCalled()
 })
})
```

Tnx for the ideas guys

**Comment by bopfer** [comment]
There is generally no reason to use `act` with react-testing-library.  It already wraps an act around the necessary functions, like `fireEvent`.

I do think switching to react-testing-library is a good idea though.  It's much nicer for testing react apps.

**Comment by julian-linux** [comment]
> 
> 
> There is generally no reason to use `act` with react-testing-library. It already wraps an act around the necessary functions, like `fireEvent`.
> 
> I do think switching to react-testing-library is a good idea though. It's much nicer for testing react apps.

if i not put `act`, it will show
```
 Attempted to log "Warning: An update to Form inside a test was not wrapped in act(...).
  ....

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
....
```

**Comment by babramczyk** [comment]
I'm also getting the error above.

My guess at this point is that, since field validation in react-form-hook is `async`, that's what causing the hanging DOM changes.

And since the `act` that RTL wraps events is `act(() => { ... })` instead of `act(async () => { ... })`, it doesn't handle this case..? Just a guess though.

**Comment by bluebill1049** [comment]
@babramczyk you are right, it's because of some of the validation/internal code can be `async`.

**Comment by babramczyk** [comment]
Out of curiosity, why are validations asynchronous?

**Comment by krnlde** [comment]
Because they can be expensive (time-consuming), especially when you have backend validation involved.

**Comment by linx8** [comment]
As @julian-linux stated, the correct way is:

```
  await act(async () => {
      fireEvent.submit(getByTestId('form'))
  })
  expect(onSubmit).toBeCalled()
```
No warning, awaits as expected!

**Comment by bluebill1049** [comment]
thanks @linx8 👍

**Comment by babramczyk** [comment]
@linx8 The bummer about that method, though, is the extra thought/code that's associated, that (if you're not doing asynchronous validation) has nothing to do with what the user is doing.

In other words, ideally when you're writing tests you should be able to write code the exactly resembles how the user is interacting with the page. If I don't have asynchronous validation, the behavior in the browser seems synchronous -- the fields are validated (in perception) immediately. But then when I hop over to reproduce that in a test, I have to remember that my third party form library has implementation details that technically doesn't behave that way, since it queues validation in the event loop, whether it's async or not.

I'm not sure of the solution though. It might just be an inherent issue with testing web pages. Though  I'm still not convinced on why validation needs to be wrapped in a promise unconditionally.

**Comment by bluebill1049** [comment]
I think this is a good opportunity to clear a few things in here why you need `async` act:

RHF is validating every rules async by design. Async validation should be the first-class citizen on validation, and I believe front end lives in an async world. At the same time by doing so, it will lead us into consistent API rather have you as a developer to worry about when to use which API (DX). Can you imagine you have to remember two methods to trigger a validation (one async and the other sync)? Not for me, I like simple and elegant, not pages and pages of API documentation (we only have one page with all the API and that's one of the values of RHF). On the other hand, it also simplifies the codebase without having to divide validation into two categories, and hence RHF is one of the smallest form validation lib out there. 

However, there are always things to improve and be informed about it. I would spend time on update the FAQ page and you can contribute too. The website is open source and everyone can help. 

Hope these makes sense.

**Comment by babramczyk** [comment]
I definitely see what you mean. At the same time, you can still `await` a non-async function, so even if validation was sometimes async and sometimes not, consuming code could still always `await` by default.

Regardless, I don't care to debate it much anymore. The nature of the web is definitely a big part of the blame, where the event loop becomes an implementation detail that sometimes throws a wrench into testing (or even production?) code.

**Comment by MauriceOppenberger** [comment]
Hi folks, I am testing my form component and, thanks to the above comments, got both tests to work as intended. However, I was hoping to get feedback on which approach is best practice and more reliable. 

First option
`
const { container, getByText, getAllByText, getByLabelText } = render(
      <Signup />
    );
    const submit = container.querySelector('button');
    const input = container.querySelector('[id=firstName]');
    await act(async () => {
     fireEvent.click(submit);
     });
     expect(getByText(/firstname is required/i)).toBeInTheDocument();
     expect(getByText(/lastname is required/i)).toBeInTheDocument();
     expect(getByText(/email is required/i)).toBeInTheDocument();
     expect(getAllByText(/password is required/i)).toHaveLength(2);

     await act(async () => {
      await fireEvent.change(input, { target: { value: 'test' } });
     });
    await act(async () => {
       fireEvent.click(submit);
    });
     expect(container.querySelector('[id=error-firstname]')).toBeNull();
     expect(getByText(/lastname is required/i)).toBeInTheDocument();
`
Second option
`
const { container, getByText, getAllByText, getByLabelText } = render(
      <Signup />
    );
    const submit = container.querySelector('button');
    const input = container.querySelector('[id=firstName]')
fireEvent.click(submit);
    await waitFor(() => {
      expect(getByText(/firstname is required/i)).toBeInTheDocument();
      expect(getByText(/lastname is required/i)).toBeInTheDocument();
      expect(getByText(/email is required/i)).toBeInTheDocument();
      expect(getAllByText(/password is required/i)).toHaveLength(2);
    });
    fireEvent.change(input, { target: { value: 'test' } });
    fireEvent.click(submit);
    await waitFor(() => {
      expect(container.querySelector('[id=error-firstname]')).toBeNull();
      expect(getByText(/lastname is required/i)).toBeInTheDocument();
    });
`
thanks

**Comment by tim-phillips** [comment]
Y'all! The user-event library is async, so we can just await it!!!

https://github.com/testing-library/user-event

```js
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

it('renders input with value', async () => {
  render(<Form />);
  const value = '10';
  const input = screen.getByLabelText(/label name/i);
  await userEvent.type(input, value);
  expect(input).toHaveValue(value);
});
```

**Comment by ohmyhusky** [comment]
In react@experimental, `mode: "onBlur",` seens doesn't trigger validation when firing blur event:
```js
    fireEvent.click(emailInput);
    fireEvent.blur(emailInput);
    await waitFor(() => getByText("Email is required.")); // TestingLibraryElementError: Unable to find an element with the text: Email is required.
   // The submit event can still trigger validation
```
It works on the browser, just not in the node testing environment.
Have anyone had the same issue in node testing with react@experimental?

**Comment by lassekl92** [comment]
For me it doesn't work to await userEvent.type. Still get the error. Wrapping in act did work though.

**Comment by bluebill1049** [comment]
we have a testing section now: https://react-hook-form.com/advanced-usage/#TestingForm probably worth to have  read

**Comment by mjbertenburg** [comment]
For me, awaiting `userEvent.type.` does not work either. Like @lassekl92 , I have to give it an extra `act` as well:

```javascript
    await act(async () => {
      await userEvent.selectOptions(screen.getByLabelText("someSelect"), "someValue")
    })
```

**Comment by mgs95** [comment]
I have a problem with this, wrapping with act en input update that calls both setValue and clearErrors functions makes the async to not finish. With just the setValue it works fine, adding the clearErrors (or setError) makes it infinite to process.

**Comment by hanvyj** [comment]
I'm having issues with react testing library, similar model to those used above:

```// Mock
      mockUseParams.mockReturnValue({ organisationId: "organisation-1" });
      mockUseSelector.mockReturnValue([
        { term: "Brute Force", count: 2000000 },
        { term: "System Information Discovery", count: 1200 },
        { term: "Exploit Public-Facing Application", count: 50 },
        { term: "Internal Spearphishing", count: 20 },
      ]);
      const onSubmit = jest.fn();

      // Setup
      render(
          <MyFormComponent
            onSubmit={onSubmit}
            onRemove={jest.fn()}
            onCancel={jest.fn()}
          />)

      act(() => {
        userEvent.click(screen.getByRole("checkbox", { name: "Something"}));
      });
      act(() => {
        fireEvent.submit(screen.getByRole("button", { name: "add" }));
      });
      await waitFor(() => screen.getByRole("button", { name: "add" }));

      // Assert - Form is submitted
      expect(onSubmit.mock.calls[0][0]).toEqual(outputExpression);
 ```
 
But it's very intermittent. When I run the test individually it works fine. Every time. Then when I run it with other tests, it fails randomly.

**Comment by jorisre** [comment]
@hanvyj Please open a new issue with a complete example in codesanxbox.io, thanks

**Comment by hanvyj** [comment]
@jorisre spent ages trying to write up and reproduce it in a complete example and could not get it to fail in the same way.

Found some odd behaviour, it worked fine in watch mode, and with `--runInBand` option with jest, or running as a single test file. Basically, didn't work in parallel. So spent ages looking for some conflicting state, trying to reproduce it with loads of test cases running in parallel.

Then found this: https://github.com/testing-library/react-testing-library/issues/848.

Basically, it was a performance issue. The `submit` was taking longer than the timeout of the `waitFor` (default is only 1s) in:

```
await waitFor(() => expect(onSubmit).toHaveBeenCalled());
```
Simply putting:
```
await waitFor(() => expect(onSubmit).toHaveBeenCalled(), { timeout: 5000 });
```
resolved the issue.

It was failing when running the full test suite in parallel simply because there was more load, presumably, so slightly slower performance. 

I've noticed quite slow performance with the `onSubmit` in the form in development mode, but seems fine in production.

So ultimately, it isn't an issue with the library, but hopefully anyone googling similar issues might stumble upon this comment and find some useful info.

**Comment by mylonasg88** [comment]
For what it's worth, those who can't make this work, suggestion above worked:

```
 await act(async () => {
    fireEvent.submit(fragment.getByTestId('form'));
     // OR
    fireEvent.submit(screen.getByRole('button', { name: /Save/i }));
});
```
However if you are struggling with expect
`expect(onSaveMock.mock.calls.length).toBe(1);`
which has result `0` instead of `1` check `errors` that could be produced in your form. 
In my case I was not populating the form, which I thought was _OK_ to skip, and submit button would never work. It would create error labels on the form.

Just a tip💡

**Comment by crisecheverria** [comment]
I have issues testing react-hook-form with TextField from Material-UI. The problem is the validation rule, if I remove it it works.

```js
<Controller
           ...
           rules={{
                required: getRequired(t),
                validate:                             // <== This is the issue. If I remove it it works
                      locale === "es-CL"
                              ? getValidation(t)
                              : getValidation2(t)
                 }}
            render={({ onChange, onBlur, value }) => (
                  <TextField
                            error={Boolean(errors?.documentNumber)}
                            helperText={errors?.documentNumber?.message}
                            ...
                  />
             )}
     />
```

But if I keep the `validate` prop then react-testing-library shows the validation error, therefore I cant submit the form. This happens only on test environment.

Test example

```js
renderWithProviders(
        <MockedProvider mocks={[mock]} addTypename={false}>
               <Register />
        </MockedProvider>
      );

      await waitFor(() => {
        userEvent.type(screen.getByTestId(document), docValue);
        userEvent.type(screen.getByTestId(pass), passValue);
      });

      const submitButton = screen.getByRole("button", {
        name: /Submit/i
      });

      await waitFor(() => fireEvent.click(submitButton));
```

**Comment by Moshyfawn** [comment]
Take a look at [this CSB](https://codesandbox.io/s/withered-firefly-ufogs?file=/src/Demo.tsx) for RHF use with [the <DateRangePicker />](https://reactnicedates.hernansartorio.com/#overview) from [react-nice-dates](https://github.com/hernansartorio/react-nice-dates).
_P.S. I've never used the lib, so there's some weirdness with it's input class styles and some issue with children, but I've used the docs example, so, should be fine. Other then that, the example should suffice._

**Comment by bluebill1049** [comment]
you can also refer to the library's testing: https://github.com/react-hook-form/react-hook-form/blob/master/src/__tests__/controller.test.tsx

**Comment by Moshyfawn** [comment]
Ooops, wrong thread, ignore me. Was referring to #6793
